# 11 - Counterfactual Calculation

In [229]:
import pandas as pd
import pickle

In [230]:
# Display all rows and columns
pd.set_option('display.max_colwidth', None)  # Show full content in each cell
pd.set_option('display.max_rows', None)      # Show all rows
pd.set_option('display.max_columns', None)   # Show all columns

In [231]:
df = pd.read_parquet("../data/processed/X_test.parquet")

In [232]:
df.head()

Position_IT Support  Position_Production Technician I  \
654                     0                                 1   
79                      0                                 1   
2775                    0                                 0   
1785                    0                                 0   
663                     0                                 1   

      Position_Area Sales Manager  Position_Production Manager  \
654                             0                            0   
79                              0                            0   
2775                            0                            0   
1785                            0                            0   
663                             0                            0   

      Position_Production Technician II  Position_Sales Manager  \
654                                   0                       0   
79                                    0                       0   
2775                                  1                       0   
1785                                  0                       0   
663                                   0                       0   

      Position_Enterprise Architect  Position_Network Engineer  \
654                               0                          0   
79                                0                          0   
2775                              0                          0   
1785                              0                          0   
663                               0                          0   

      Position_Sr. Network Engineer  Position_Database Administrator  \
654                               0                                0   
79                                0                                0   
2775                              0                                0   
1785                              0                                0   
663                               0                                0   

      Position_Data Analyst  Position_Software Engineer  Position_Sr. DBA  \
654                       0                           0                 0   
79                        0                           0                 0   
2775                      0                           0                 0   
1785                      0                           0                 0   
663                       0                           0                 0   

      Position_Sr. Accountant  Position_Administrative Assistant  \
654                         0                                  0   
79                          0                                  0   
2775                        0                                  0   
1785                        0                                  0   
663                         0                                  0   

      Position_Accountant I  Position_Shared Services Manager  \
654                       0                                 0   
79                        0                                 0   
2775                      0                                 0   
1785                      0                                 0   
663                       0                                 0   

      Position_IT Director  Position_CIO  Position_Principal Data Architect  \
654                      0             0                                  0   
79                       0             0                                  0   
2775                     0             0                                  0   
1785                     0             0                                  0   
663                      0             0                                  0   

      Position_IT Manager - DB  Position_IT Manager - Support  \
654                          0                              0   
79                           0                              0   
2775                         0                              0   

In [233]:
# Load the model
model_path = "../models/xgb_model.pkl"
with open(model_path, 'rb') as file:
    xgb_model = pickle.load(file)

In [234]:
probs = xgb_model.predict_proba(df)[:, 1]  # Get the probability of the positive class

In [235]:
# Append probabilities as a new column in the dataframe
df['probability'] = probs

In [236]:
# Define thresholds for filtering
threshold_lower = 0.45
threshold_upper = 0.55

In [237]:
# Filter rows based on probabilities being close to the threshold range (e.g., 0.45 to 0.5)
filtered_df = df[(df['probability'] >= threshold_lower) & (df['probability'] <= threshold_upper)]

In [238]:
df.YearsExperience.value_counts()

YearsExperience
13    77
10    70
11    51
12    40
14    24
9     18
8     14
16     8
7      4
15     3
17     3
19     1
Name: count, dtype: int64

In [239]:
filtered_df.head()

Position_IT Support  Position_Production Technician I  \
2146                    0                                 0   
2861                    0                                 1   
2951                    0                                 0   
2867                    0                                 1   
2873                    0                                 1   

      Position_Area Sales Manager  Position_Production Manager  \
2146                            0                            0   
2861                            0                            0   
2951                            0                            0   
2867                            0                            0   
2873                            0                            0   

      Position_Production Technician II  Position_Sales Manager  \
2146                                  1                       0   
2861                                  0                       0   
2951                                  0                       0   
2867                                  0                       0   
2873                                  0                       0   

      Position_Enterprise Architect  Position_Network Engineer  \
2146                              0                          0   
2861                              0                          0   
2951                              0                          0   
2867                              0                          0   
2873                              0                          0   

      Position_Sr. Network Engineer  Position_Database Administrator  \
2146                              0                                0   
2861                              0                                0   
2951                              0                                0   
2867                              0                                0   
2873                              0                                0   

      Position_Data Analyst  Position_Software Engineer  Position_Sr. DBA  \
2146                      0                           0                 0   
2861                      0                           0                 0   
2951                      1                           0                 0   
2867                      0                           0                 0   
2873                      0                           0                 0   

      Position_Sr. Accountant  Position_Administrative Assistant  \
2146                        0                                  0   
2861                        0                                  0   
2951                        0                                  0   
2867                        0                                  0   
2873                        0                                  0   

      Position_Accountant I  Position_Shared Services Manager  \
2146                      0                                 0   
2861                      0                                 0   
2951                      0                                 0   
2867                      0                                 0   
2873                      0                                 0   

      Position_IT Director  Position_CIO  Position_Principal Data Architect  \
2146                     0             0                                  0   
2861                     0             0                                  0   
2951                     0             0                                  0   
2867                     0             0                                  0   
2873                     0             0                                  0   

      Position_IT Manager - DB  Position_IT Manager - Support  \
2146                         0                              0   
2861                         0                              0   
2951                         0                              0   

In [240]:
row_to_modify = filtered_df.iloc[1].copy()

In [241]:
# Display the original row and its probability
print("Original Row:")
print(row_to_modify)

Original Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000000
Position_Sr. Accountant              

In [242]:
# Modify the row: change RaceDesc_White from 1 to 0 and RaceDesc_Asian from 0 to 1
if 'RaceDesc_White' in row_to_modify.index and 'RaceDesc_Asian' in row_to_modify.index:
    row_to_modify['RaceDesc_White'] = 0
    row_to_modify['RaceDesc_Asian'] = 1

In [243]:
# Display the modified row
print("\nModified Row:")
print(row_to_modify)


Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000000
Position_Sr. Accountant             

In [244]:
# Convert the row back into a DataFrame (model expects a DataFrame for predictions)
modified_row_df = row_to_modify.to_frame().T
probability = modified_row_df.probability
probability

2861    0.488037
Name: probability, dtype: float64

In [245]:
modified_row_df.drop(columns="probability", inplace=True)

In [246]:
# Recalculate the probability for the modified row
counterfactual_probability = xgb_model.predict_proba(modified_row_df)[:, 1][0]

# Display the counterfactual probability
print("\nCounterfactual Probability:")
print(counterfactual_probability)


Counterfactual Probability:
0.99927


The counterfactual probability of **0.99927** shows that changing `RaceDesc_White` to 0 and `RaceDesc_Asian` to 1 caused the model to predict the positive class with almost 100% confidence. This suggests the model might be overly sensitive to these features, possibly due to their high importance or correlations in the training data.

## **Modifiable Attributes : gender, age, race**

In [247]:
def modify_and_recalculate_probability(row: pd.Series, 
                                      xgb_model: object, 
                                      attribute: str, 
                                      new_value: int | str) -> dict:
    """
    Modify a single attribute of a selected row, recalculate probability, and return results.

    Parameters:
    row (pd.Series): The original row selected by the user.
    xgb_model: The trained XGBoost model for predicting probabilities.
    attribute (str): The attribute to modify ('Sex', 'Race', or 'YearsExperience').
    new_value: The new value to set for the selected attribute.
        - For 'Sex': 0 (Female) or 1 (Male)
        - For 'Race': 'White', 'Black', or 'Asian'
        - For 'YearsExperience': 7, 13, or 17

    Returns:
    dict: A dictionary containing the original probability, modified row, and counterfactual probability.
    """
    import pandas as pd

    # Validate inputs
    if attribute not in ['Sex', 'Race', 'YearsExperience']:
        raise ValueError("Invalid attribute. Must be 'Sex', 'Race', or 'YearsExperience'.")

    if attribute == 'Sex' and new_value not in [0, 1]:
        raise ValueError("Invalid value for 'Sex'. Must be 0 (Female) or 1 (Male).")

    if attribute == 'Race' and new_value not in ['White', 'Black', 'Asian']:
        raise ValueError("Invalid value for 'Race'. Must be 'White', 'Black', or 'Asian")

    if attribute == 'YearsExperience' and new_value not in [7, 13, 17]:
        raise ValueError("Invalid value for 'YearsExperience'. Must be 7, 13, or 17.")

    # Copy the row to avoid modifying the original data
    modified_row = row.copy()

    # Save the original probability
    original_probability = modified_row.get('probability', None)
    if original_probability is None:
        raise KeyError("The column 'probability' is missing in the row.")

    # Modify the specified attribute
    if attribute == 'Sex':
        modified_row['Sex'] = new_value
    elif attribute == 'Race':
        # Dynamically identify all columns that start with 'RaceDesc_'
        race_columns = [col for col in modified_row.index if col.startswith('RaceDesc_')]
        # Reset all race columns to 0
        for col in race_columns:
            modified_row[col] = 0

        # Set the selected race column to 1
        race_column_mapping = {
            'White': 'RaceDesc_White',
            'Black': 'RaceDesc_Black or African American',
            'Asian': 'RaceDesc_Asian'
        }
        race_col = race_column_mapping.get(new_value)
        if race_col and race_col in modified_row:
            modified_row[race_col] = 1
    elif attribute == 'YearsExperience':
        modified_row['YearsExperience'] = new_value

    # Convert the modified row to a DataFrame (XGBoost model expects this format)
    modified_row_df = modified_row.to_frame().T
    modified_row_df.drop(columns="probability", inplace=True)

    # Recalculate probability for the modified row
    counterfactual_probability = xgb_model.predict_proba(modified_row_df)[:, 1][0]

    return {
        "original_probability": original_probability,
        "counterfactual_probability": counterfactual_probability,
        "modified_row": modified_row
    }

In [248]:
result = modify_and_recalculate_probability(row_to_modify, xgb_model, 'Race', 'Asian')
print(result)

{'original_probability': np.float64(0.4880373179912567), 'counterfactual_probability': np.float32(0.99927), 'modified_row': Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000


In [249]:
# Filter for a row where 'RaceDesc_White' is 1
white_row = df[df['RaceDesc_White'] == 1].iloc[0]

# Change race from 'White' to 'Black'
result = modify_and_recalculate_probability(white_row, xgb_model, 'Race', 'Black')

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])

Original Probability: 0.6975823640823364
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               1.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

In [250]:
# Filter for a row where 'Sex' is 1 (Male)
male_row = filtered_df[filtered_df['Sex'] == 1].iloc[0]

# Change gender from Male to Female
result = modify_and_recalculate_probability(male_row, xgb_model, 'Sex', 0)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])

Original Probability: 0.46454301476478577
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           1.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.00

In [255]:
# Filter for a row with a specific value for 'YearsExperience' (e.g., 7, 13, or 17)
years_row = df[df['YearsExperience'] == 7].iloc[1]

# Change 'YearsExperience' to a new value (e.g., 13)
result = modify_and_recalculate_probability(years_row, xgb_model, 'YearsExperience', 17)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])


Original Probability: 0.7582640051841736
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

In [256]:
# Filter for a row with a specific value for 'YearsExperience' (e.g., 7, 13, or 17)
years_row = filtered_df[filtered_df['YearsExperience'] == 13].iloc[0]

# Change 'YearsExperience' to a new value (e.g., 17)
result = modify_and_recalculate_probability(years_row, xgb_model, 'YearsExperience', 17)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])

Original Probability: 0.5453081727027893
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               1.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

By changing individual attributes like race, gender, or years of experience, we observed significant shifts in the model's predictions. These results highlight the sensitivity of the model to specific features and emphasize the need to evaluate feature importance and potential biases. This demonstrates the impact of single-value modifications on prediction outcomes.

In [260]:
# Classification threshold (e.g., 0.5)
classification_threshold = 0.5

candidates = []

# Define attributes and values to modify
attributes = [
    {"attribute": "Sex", "values": [0, 1]},  # Female to Male or vice versa
    {"attribute": "Race", "values": ["White", "Black", "Asian"]},  # Race changes (mapped to columns)
    {"attribute": "YearsExperience", "values": [7, 13, 17]}  # Experience levels
]

# Mapping for race columns (one-hot encoded)
race_column_mapping = {
    "White": "RaceDesc_White",
    "Black": "RaceDesc_Black or African American",
    "Asian": "RaceDesc_Asian"
}

# Iterate through rows to identify significant changes
for idx, row in df.iterrows():
    for attr in attributes:
        if attr["attribute"] == "Race":
            # Handle one-hot encoding for race
            for race, race_col in race_column_mapping.items():
                if race_col in row and row[race_col] == 1:  # Skip if current race matches target race
                    continue

                # Create a modified copy of the row for race modification
                modified_row = row.copy()

                # Reset all race columns to 0
                for col in race_column_mapping.values():
                    if col in modified_row:
                        modified_row[col] = 0

                # Set the target race column to 1
                if race_col in modified_row:
                    modified_row[race_col] = 1

                # Apply modification and calculate probabilities
                result = modify_and_recalculate_probability(modified_row, xgb_model, "Race", race)
                original_prob = result["original_probability"]
                counterfactual_prob = result["counterfactual_probability"]

                # Identify the original race
                original_race = next(
                    (k for k, v in race_column_mapping.items() if v in row and row[v] == 1),
                    "Unknown"
                )

                # Check if the change crosses the classification threshold
                if (original_prob < classification_threshold <= counterfactual_prob) or \
                   (original_prob >= classification_threshold > counterfactual_prob):
                    candidates.append({
                        "row_index": idx,
                        "attribute": "Race",
                        "original_value": original_race,
                        "new_value": race,
                        "original_probability": original_prob,
                        "counterfactual_probability": counterfactual_prob,
                        "probability_change": abs(counterfactual_prob - original_prob)
                    })
        else:
            # Handle other attributes (Sex, YearsExperience)
            for value in attr["values"]:
                if row[attr["attribute"]] == value:  # Skip if the current value matches the new value
                    continue

                # Apply modification and calculate probabilities
                result = modify_and_recalculate_probability(row, xgb_model, attr["attribute"], value)
                original_prob = result["original_probability"]
                counterfactual_prob = result["counterfactual_probability"]

                # Check if the change crosses the classification threshold
                if (original_prob < classification_threshold <= counterfactual_prob) or \
                   (original_prob >= classification_threshold > counterfactual_prob):
                    candidates.append({
                        "row_index": idx,
                        "attribute": attr["attribute"],
                        "original_value": row[attr["attribute"]],
                        "new_value": value,
                        "original_probability": original_prob,
                        "counterfactual_probability": counterfactual_prob,
                        "probability_change": abs(counterfactual_prob - original_prob)
                    })

In [ ]:
# Sort by probability change for better insights
candidates_df = pd.DataFrame(candidates)
candidates_df = candidates_df.sort_values(by="probability_change", ascending=False)
candidates_df = candidates_df.reset_index(drop=True)

In [263]:
# Define custom sort order for the attribute column
attribute_order = {"Sex": 0, "Race": 1, "YearsExperience": 2}

# Sort DataFrame
candidates_df = candidates_df.sort_values(
    by=["attribute", "original_probability", "probability_change"],
    key=lambda col: col.map(attribute_order) if col.name == "attribute" else col,
    ascending=[True, True, False]  # Ascending for attribute and original_probability, descending for probability_change
)
candidates_df.reset_index(drop=True, inplace=True)
candidates_df

row_index        attribute original_value new_value  \
0         1939              Sex            0.0         1   
1         2880              Sex            0.0         1   
2          323              Sex            1.0         0   
3         1572              Sex            1.0         0   
4         1716              Sex            1.0         0   
5         2762              Sex            0.0         1   
6         2541              Sex            0.0         1   
7          941              Sex            0.0         1   
8         1023              Sex            0.0         1   
9         2063              Sex            0.0         1   
10        1337              Sex            0.0         1   
11        2221              Sex            1.0         0   
12        2728              Sex            0.0         1   
13        1844              Sex            0.0         1   
14        3084              Sex            0.0         1   
15         985              Sex            1.0         0   
16         570              Sex            0.0         1   
17        1187              Sex            0.0         1   
18         925              Sex            0.0         1   
19          71              Sex            0.0         1   
20         336              Sex            0.0         1   
21         304              Sex            0.0         1   
22        2516              Sex            0.0         1   
23        3093              Sex            0.0         1   
24        1091              Sex            1.0         0   
25        1240              Sex            1.0         0   
26         995              Sex            0.0         1   
27         983              Sex            1.0         0   
28         576              Sex            0.0         1   
29         444              Sex            0.0         1   
30        2867              Sex            0.0         1   
31        1391              Sex            1.0         0   
32        2951              Sex            1.0         0   
33        2861              Sex            0.0         1   
34        3184              Sex            0.0         1   
35        1805              Sex            1.0         0   
36        1478              Sex            0.0         1   
37          49              Sex            0.0         1   
38        2284              Sex            1.0         0   
39        2704              Sex            0.0         1   
40        1130              Sex            1.0         0   
41        2709              Sex            0.0         1   
42        2898              Sex            1.0         0   
43        2530              Sex            1.0         0   
44         876              Sex            1.0         0   
45        2819              Sex            0.0         1   
46        2477              Sex            0.0         1   
47        2797              Sex            0.0         1   
48        2542              Sex            1.0         0   
49         694              Sex            0.0         1   
50        1152              Sex            0.0         1   
51        1628              Sex            0.0         1   
52        1318              Sex            0.0         1   
53        2584              Sex            0.0         1   
54         677              Sex            0.0         1   
55        1939             Race          White     Black   
56        1939             Race          White     Asian   
57         323             Race          White     Black   
58         323             Race          White     Asian   
59        1393             Race          White     Black   
60        1393             Race          White     Asian   
61        1572             Race          White     Asian   
62        2762             Race          Asian     White   
63        2541             Race          Black     Asian   
64        2541             Race          Black     White   
65         941             Race          Whit

In [266]:
# Group by 'row_index' and count occurrences
duplicate_candidates = candidates_df.groupby("row_index").size()

# Filter for row indices that appear more than once
duplicate_row_indices = duplicate_candidates[duplicate_candidates > 1].index

# Filter original DataFrame for these row indices and select required columns
duplicates_df = candidates_df[candidates_df["row_index"].isin(duplicate_row_indices)][
    ["row_index", "attribute", "original_value", "new_value", "original_probability", "counterfactual_probability", "probability_change"]
]
duplicates_df.reset_index(drop=True, inplace=True)
duplicates_df.sort_values(by="row_index")

row_index        attribute original_value new_value  \
37          49              Sex            0.0         1   
114         49             Race          White     Asian   
72          71             Race          White     Asian   
186         71  YearsExperience           13.0        17   
19          71              Sex            0.0         1   
273         87  YearsExperience           13.0         7   
140         87             Race        Unknown     White   
122        107             Race          White     Black   
250        107  YearsExperience           11.0         7   
251        107  YearsExperience           11.0        13   
128        186             Race        Unknown     Black   
258        186  YearsExperience           11.0         7   
62         287             Race          White     Asian   
63         287             Race          White     Black   
210        300  YearsExperience           13.0        17   
93         300             Race          White     Asian   
92         300             Race          White     Black   
209        300  YearsExperience           13.0         7   
21         304              Sex            0.0         1   
190        304  YearsExperience           12.0         7   
189        304  YearsExperience           12.0        17   
191        304  YearsExperience           12.0        13   
44         323             Race          White     Asian   
148        323  YearsExperience            8.0        13   
147        323  YearsExperience            8.0        17   
2          323              Sex            1.0         0   
43         323             Race          White     Black   
20         336              Sex            0.0         1   
188        336  YearsExperience           11.0        13   
187        336  YearsExperience           11.0        17   
73         336             Race          Black     Asian   
229        347  YearsExperience           10.0         7   
105        347             Race          White     Black   
29         444              Sex            0.0         1   
96         444             Race          White     Black   
97         444             Race          White     Asian   
214        444  YearsExperience           11.0        13   
213        444  YearsExperience           11.0        17   
170        552  YearsExperience            9.0        13   
169        552  YearsExperience            9.0        17   
180        570  YearsExperience            9.0        17   
181        570  YearsExperience            9.0        13   
16         570              Sex            0.0         1   
243        574  YearsExperience           14.0        13   
116        574             Race        Unknown     Black   
242        574  YearsExperience           14.0         7   
94         576             Race        Unknown     Asian   
95         576             Race        Unknown     White   
212        576  YearsExperience           11.0        13   
211        576  YearsExperience           11.0        17   
28         576              Sex            0.0         1   
265        583  YearsExperience           10.0        13   
264        583  YearsExperience           10.0         7   
90         670             Race          White     Asian   
207        670  YearsExperience           11.0        17   
208        670  YearsExperience           11.0        13   
91         670             Race          White     Black   
278        696  YearsExperience           17.0        13   
277        696  YearsExperience           17.0         7   
39         876              Sex            1.0         0   
259        876  YearsExperience           13.0         7   
142        914             Race        Unknown     Asian   
141        914             Race        Unknown     Black   
279        914  YearsExperience           13.0         7   
71         925             Race          Black     Asian   
183        925  YearsExperience           12.

**Note:** We can use this table to identify some interesting candidates for the MVP, as it highlights rows where multiple attributes significantly impact the classification threshold.

Examples: 323, 925, 1023

In [275]:
row_indices = [323, 925, 1023]  # Indices to select
selected_rows = df.loc[row_indices]

In [277]:
selected_rows.drop(columns="probability", inplace=True)

In [ ]:
# Save the selected rows to a new parquet file without index
selected_rows.to_parquet("../data/app/static_data.parquet", index=False)
selected_rows